src/rust_connection/mod.rs

```diff
@@ -659,21 +659,33 @@ impl<S: Stream> Connection for RustConnection<S> {
     }
 
     fn generate_id(&self) -> Result<u32, ReplyOrIdError> {
-        let mut id_allocator = self.id_allocator.lock().unwrap();
-        if let Some(id) = id_allocator.generate_id() {
+        let mut inner = self.inner.lock().unwrap();
+        if let Some(id) = inner.inner.generate_id() {
             Ok(id)
         } else {
             use crate::protocol::xc_misc::{self, ConnectionExt as _};
 
+            // n.b. notgull: temporarily drop inner lock so that extension_information
+            // and xc_misc_get_xid_range can run without deadlocking
+            // could be improved by, e.g, passing the display lock to the
+            // extension_information() function and manually sending the GetXidRange
+            // request. Re-evaluate this part of the code once x11rb_protocol
+            // is in a good place
+            drop(inner);
```

Comment:
Isn't that racy? Another thread may end up calling this function in parallel and also do an `get_xid_range` request. I would expect both replies to be the same, which means that:

- Thread 1 updates the id allocator, grabs the first free ID
- Thread 2 does the same, effectively re-setting the id allocator and getting the *same* ID

This would then cause an ID collision down the road. (I guess `IDChoice` errors? I am not actually sure.)

Why are you even doing this move? Why not keep the fields as they were with the extra Mutex for the ID allocator?

Reply:
> Why are you even doing this move? Why not keep the fields as they were with the extra Mutex for the ID allocator?

The general idea is that ID allocation is a part of the protocol, so it should go in `Connection` to reduce complexity on the I/O side. But now that I see how it turns out, yeah I see what you mean.